### ISSUE TITLE
'${' autosuggestion and completions causing bytes to leak out.

### ISSUE THREAD

**Issue by floam** [fish-shell/fish-shell#2796]
On my system, even with my entire ~/.local/share/fish and ~/.config/fish pushed out of the way, almost no environment, something is happening if I type ${ - immediately autosuggest has a problem: 

<img width="178" alt="screenshot 2016-03-05 at 6 37 50 am" src="https://cloud.githubusercontent.com/assets/291142/13548111/efed2b04-e29c-11e5-8031-5c50bf7df396.png">

and the completion/pager if I arrow down:
<img width="489" alt="screenshot 2016-03-05 at 6 38 09 am" src="https://cloud.githubusercontent.com/assets/291142/13548118/07688206-e29d-11e5-8cb7-64283e8f191c.png">

If I hit right arrow to let it finish, completing, then remove the syntax from the front, it seems it's really just ${ there, same if I echo it to terminal, or into a to a file and inspect that.

<img width="254" alt="screenshot 2016-03-05 at 6 42 58 am" src="https://cloud.githubusercontent.com/assets/291142/13548134/93fd8de2-e29d-11e5-9a18-bbff2b9b8e21.png">
<img width="178" alt="screenshot 2016-03-05 at 6 41 21 am" src="https://cloud.githubusercontent.com/assets/291142/13548141/9f2c0734-e29d-11e5-8050-db9f2eba1ce1.png">

If I paste the output somewhere it appears some stuff and nulls:
```
0000000          ef92b7ef        000094b7                                
0000006
```

**Comment by floam** [comment]
And in history now: 
<img width="137" alt="screenshot 2016-03-05 at 6 53 32 am" src="https://cloud.githubusercontent.com/assets/291142/13548180/0b9ecaa4-e29f-11e5-919a-434d4e0332a4.png">

**Comment by faho** [comment]
Hummm.... my guess is you've got this in your history, possibly imported from bash.

I can't reproduce it - can you try again with your history file moved out of the way (and possibly upload that if it's working then)?

**Comment by floam** [comment]
That's what I thought - moving that away was the first thing I did. It's really not the history: 
<img width="396" alt="screenshot 2016-03-05 at 7 07 21 am" src="https://cloud.githubusercontent.com/assets/291142/13548272/06cc3a0a-e2a1-11e5-9087-c35ee5ae2fb2.png">

**Comment by faho** [comment]
Oh, I managed to reproduce - not sure what went wrong the first time.

These characters are 0xFDD0u and 0xFDD2u, two unicode "noncharacters". The first one we use as "RESERVED_CHAR_BASE" since #2713. Seems we are leaking internal details. @krader1961: Mind taking a look?

**Comment by floam** [comment]
Nothing special about the braces - most non-alphanumeric characters after $ do it in fact.

**Comment by floam** [comment]
It's going to be an older build than #2713 that introduced this. I have one sitting around from Feb 28:

<img width="367" alt="screenshot 2016-03-05 at 9 58 07 am" src="https://cloud.githubusercontent.com/assets/291142/13549395/e08ecbe2-e2b8-11e5-939e-76be3fddcc56.png">

They're different characters from before the change, it's getting glyphs from my Powerline font. With another font it looks exactly the same, except it's only on suggestions and not completions as @ridiculousfish just added completions on $

**Comment by oranja** [comment]
I ran out of time while I stared at the code (very unfamiliar with the codebase, to be honest), so I can't confirm my findings yet, but I think I have a suspect for your consideration:
In [completer::try_complete_variable], `variable_start` is a `size_t`, but it's assigned with `wcstring::npos`, which is defined as *-1* (or directly -1, a few lines below).
`size_t` is unsigned, as you all know well, and so at the bottom, I suspect that the test:
`variable_start != wcstring::npos` passes even when `variable_start` is intended to be `wcstring::npos`, resulting in an attempt to complete illegal variable names, such as `${`.

I admit that it still doesn't necessarily explain why the completer suggests the internal private-use variables. Unless... they are stored in one of the variable scopes (global unexported would be my wild guess), and until now no-one noticed because you had to start the variable with some alphanumeric character, or something like that...

Last minute update:
Now I see this: [6cb48c6]. Look at the change in complete.cpp - lines 1559/1561.

[completer::try_complete_variable]: https://github.com/fish-shell/fish-shell/blob/master/src/complete.cpp#L1512
[6cb48c6]: https://github.com/fish-shell/fish-shell/commit/6cb48c63800604e40ba759763ab410beceb74b89

**Comment by floam** [comment]
>Last minute update:
>Now I see this: 6cb48c6. Look at the change in complete.cpp - lines 1559/1561.

Close, it's actually the enums in this commit: https://github.com/fish-shell/fish-shell/commit/992dfcc4cec760f5fe704a7ac38478beb19f4f04

**Comment by ridiculousfish** [comment]
D'oh

**Comment by floam** [comment]
Er, nope that commit was the last good. one.I didn't make clean.

**Comment by krader1961** [comment]
`git bisect` shows this change in behavior was introduced by @ridiculousfish's commit 992dfcc4cec760f5fe704a7ac38478beb19f4f04. What's interesting is fixing the definition of **EXPAND_RESERVED_END** (it had one too many zeros) after that commit, but before any other commit, fixes the problem. However, my equivalent fix for that symbol, which also moved the range of those symbols, in commit f2246dfb343bea19beb176fb2cc534f85513b2eb doesn't have the same salutary effect. Also odd is reverting @ridiculousfish's change against git HEAD doesn't fix the behavior. There's clearly some sort of boundary violation, possibly involving unsigned chars. @oranja might be on the right track but I haven't yet looked at that part of the code.

**Comment by krader1961** [comment]
On my system the weird characters at the end of the completion are 0xFDD2 and 0xFDD4 rather than the 0xFDD0 and 0xFDD2 reported by @faho. The values I see make more sense since 0xFDD2 == **VARIABLE_EXPAND** and 0xFDD4 == **BRACKET_BEGIN** corresponding to the dollar-sign and left-brace in @floam's example. So it seems likely there is a piece of code that I overlooked in my change to limit the range of Unicode private-use characters that is using a different range. Which is exactly why I introduced my change. There were too many places in the code which were using "magic" constants rather than symbols that were themselves verified to have reasonable values. Apparently there is still at least one such place in the code that needs fixing.

**Comment by floam** [comment]
Mine were also always only 0xFDD2 and 0xFDD4.

On Sat, Mar 05, 2016 at 7:20 PM Kurtis Rader

<
mailto:Kurtis Rader <notifications@github.com>
> wrote:

On my system the weird characters at the end of the completion are 0xFDD2 and 0xFDD4 rather than the 0xFDD0 and 0xFDD2 reported by
https://github.com/faho
. The values I see make more sense since 0xFDD2 ==

VARIABLE_EXPAND

and 0xFDD4 ==

BRACKET_BEGIN

corresponding to the dollar-sign and left-brace in
https://github.com/floam
's example. So it seems likely there is a piece of code that I overlooked in my change to limit the range of Unicode private-use characters that is using a different range. Which is exactly why I introduced my change. There were too many places in the code which were using "magic" constants rather than symbols that were themselves verified to have reasonable values. Apparently there is still at least one such place in the code that needs fixing.

**Comment by floam** [comment]
Ah, but with those other non-alphanums I only ever get just <U+F002> and a correctly rendering version of whatever that symbol is, not two different /UCHARs.

**Comment by krader1961** [comment]
@floam, that's because the parser only treats a few characters, like dollar-sign and left-brace, specially. So, for example, if you enter `$!` you'll see the 0xFDD2 (**VARIABLE_EXPAND**) private-use char followed by a simple ASCII exclamation mark.

And this problem is the poster-child for why someone needs to be willing to spend a couple of hundred hours completely re-architecting how the fish parser interacts with the rest of the fish code :-)

FWIW, I don't think this needs to delay the upcoming point release. It's a pretty obscure problem.

**Comment by ridiculousfish** [comment]
My bisect shows c4d0612e5c0fbc0dd118ade84c9be055e0f2ea89

We should definitely fix this.

**Comment by floam** [comment]
> git bisect shows this change in behavior was introduced by @ridiculousfish's commit 992dfcc. 

I thought I had found the bad commit at 992dfcc earlier tonight. I could swear it was messed up and I had done a make clean and everything. But when I ran git bisect it proved me wrong and I corrected myself. 

Can you recheck? I'm doing so as that's an odd coincidence.

**Comment by floam** [comment]
I've now got a version 2.2.0-584-g992dfcc binary that is showing no symptoms.

**Comment by krader1961** [comment]
@ridiculousfish you appear to be correct that commit c4d0612e5c0fbc0dd118ade84c9be055e0f2ea89, rather than 992dfcc4cec760f5fe704a7ac38478beb19f4f04, is at fault. Which is weird since I could earlier reproducibly reproduce the problem with the latter change being my current repo head even outside of git bisect. I must have done something wrong with my `make fish; ./fish` testing. And yet I don't think that's the whole story because I could swear I reliably fixed the problem by correcting the definition of **EXPAND_RESERVED_END** (removing the extra zero) while commit 992dfcc4cec760f5fe704a7ac38478beb19f4f04 was checked out.

Despite what I wrote above commit https://github.com/fish-shell/fish-shell/commit/c4d0612e5c0fbc0dd118ade84c9be055e0f2ea89 seems like a more likely candidate for the source of this issue.

**Comment by ridiculousfish** [comment]
The issue is that `expand_string` now leaves completions in its output array, even if it returns an error. Previously it would only return results on success. The weird characters are intermediate things (primarily INTERNAL_SEPARATOR) that appear midway through expansions.

**Comment by ridiculousfish** [comment]
Fixed as 6bee85fefa968a632287d5830be6023b90b88068 . Thanks to everyone
